### DISCUSSION TITLE
SIngleton instance

### DISCUSSION THREAD

**Discussion by M4RC0Sx** [ProSavage/FactionsX#3] [Q&A] [answered]
Hi there, I have been looking through the code of FactionsX and [SavageFramework](https://github.com/SavageLabs/SavageFramework/blob/master/src/main/java/net/prosavage/baseplugin/SavagePlugin.java).

On SavagePlugin, you initialize the Singleton instance of the base plugin. 

```java
    private static SavagePlugin instance;

    public static SavagePlugin getInstance() {
        return instance;
    }

    @Override
    public void onEnable() {
        getLogger().info("Running Base Framework Enable...");
        instance = this;
```

Then, in FactionsX main class, you make this:

```java
    companion object {
        lateinit var instance: FactionsX
    }

 @ExperimentalTime
    override fun onEnable() {
        instance = this
        super.onEnable()
```

You create a new singleton instance on the FactionsX companion object and then you call super.onEnable() (of SavageFramework BasePlugin). Why is that done? Is it OK? How does it work?

Thank you!

**Comment by prosavage** [comment, marked as answer]
SavageFramework is a toolkit to build plugins on top of.
There are some setup tasks that SavageFramework runs, like setting the instance inside of the SavagePlugin class.

I also set the instance of the companion object to use the FactionsX class's functions.
So we could be calling `SavagePlugin.getInstance()` but the type for that would be `SavagePlugin` not `FactionsX`.
We set the instance in SavagePlugin so it can reference that instance in the framework's utility functions like `SavagePlugin#registerListeners`. So we need that `super.onEnable()` just so the framework can do it's own things.

**Comment by M4RC0Sx** [comment]
Nice! So if you want to access SavagePlugin instance you get it with FactionsX.getInstance() and if you want to access FactionsX instance, you get it with FactionsX.instance

**Reply by prosavage** [reply]
You don't want to do that, remember the `FactionsX` class extends `SavagePlugin` so you can just access those functions by using `FactionsX.instance`, the SavagePlugin functions can be called directly from the plugin's main class as they inhereit them.

So basically you can do all the stuff you want with the FactionsX class and ignore SavagePlugin.

**Reply by M4RC0Sx** [reply]
Nice! Thank you!
